dev/src/index.ts

```diff
@@ -987,12 +994,39 @@ export class Firestore {
     );
   }
 
+  /**
+   * Registers the requestTag associated with the listener. This is used to
+   * ensure that all listeners are detached when terminate() is called.
+   *
+   * @private
+   * @param requestTag A unique client-assigned identifier
+   */
+  registerListener(): void {
+    this.registeredListenersCount += 1;
+  }
+
+  /**
+   * Unregisters the requestTag associated with the listener. This is used to
+   * ensure that all listeners are detached when terminate() is called.
+   *
+   * @private
+   * @param requestTag A unique client-assigned identifier
+   */
+  unregisterListener(): void {
+    this.registeredListenersCount -= 1;
+  }
+
   /**
    * Terminates the Firestore client and closes all open streams.
    *
    * @return A Promise that resolves when the client is terminated.
    */
   terminate(): Promise<void> {
+    if (this.registeredListenersCount > 0) {
+      return Promise.reject(
+        'All listeners must be disconnected before terminating the client.'
```

Comment:
Please use the terms we use in our API: "onSnapshot() listeners" and "unsubscribe".